hudi-common/src/main/java/org/apache/hudi/common/table/read/SortedKeyBasedFileGroupRecordBuffer.java

```diff
@@ -0,0 +1,100 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.common.table.read;
+
+import org.apache.hudi.common.config.RecordMergeMode;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+/**
+ * An extension of {@link KeyBasedFileGroupRecordBuffer} that sorts the log records based on the record key when joining with the base file records.
+ * This assumes that the base file records are already sorted by the record key.
+ */
+public class SortedKeyBasedFileGroupRecordBuffer<T> extends KeyBasedFileGroupRecordBuffer<T> {
+
+  // when sorting is enabled, this will hold the base file record if it was not used in the previous iteration
+  private Option<T> queuedBaseFileRecord = Option.empty();
+  private Queue<String> logRecordKeysSorted = null;
+
+  public SortedKeyBasedFileGroupRecordBuffer(HoodieReaderContext<T> readerContext,
+                                             HoodieTableMetaClient hoodieTableMetaClient,
+                                             RecordMergeMode recordMergeMode,
+                                             TypedProperties props,
+                                             HoodieReadStats readStats,
+                                             Option<String> orderingFieldName,
+                                             boolean emitDelete) {
+    super(readerContext, hoodieTableMetaClient, recordMergeMode, props, readStats, orderingFieldName, emitDelete);
+  }
+
+  @Override
+  protected void initializeLogRecordIterator() {
+    logRecordIterator = records.values().stream().sorted(Comparator.comparing(BufferedRecord::getRecordKey)).iterator();
+  }
+
+  @Override
+  protected boolean hasNextBaseRecord(T baseRecord) throws IOException {
+    String recordKey = readerContext.getRecordKey(baseRecord, readerSchema);
+    int comparison;
+    while (!getLogRecordKeysSorted().isEmpty() && (comparison = getLogRecordKeysSorted().peek().compareTo(recordKey)) <= 0) {
+      String nextLogRecordKey = getLogRecordKeysSorted().poll();
+      if (comparison < 0) {
+        BufferedRecord<T> nextLogRecord = records.remove(nextLogRecordKey);
+        if (!nextLogRecord.isDelete() || emitDelete) {
+          nextRecord = nextLogRecord.getRecord();
```

Comment:
can we add java docs here. 
mainly to callout, here we are handling insert records in log files which are not present in base files.